File: tests/unit/core/dotnet/DearStory.Core.Tests/ArgumentPatchValidatorTests.cs
using System.Linq;
using System.Text.Json.Nodes;
using DearStory.Core.Schemas;
using Xunit;

namespace DearStory.Core.Tests;

public sealed class ArgumentPatchValidatorTests
{
    [Fact]
    public void Apply_rejects_unknown_enum_value()
    {
        var schema = ArgumentSchema.Parse(
            """
            {
              "type": "object",
              "properties": {
                "size": {
                  "type": "string",
                  "enum": ["small", "medium", "large"],
                  "x-dearstory-control": "radio"
                }
              }
            }
            """);

        var result = ArgumentPatchValidator.Apply(
            schema,
            JsonNode.Parse("""{"size":"medium"}""")!,
            JsonNode.Parse("""{"size":"giant"}""")!);

        Assert.False(result.Accepted);
        Assert.Equal("args.enum", result.Diagnostics.Single().Code);
        Assert.Equal("medium", result.UpdatedArguments["size"]!.GetValue<string>());
    }

    [Fact]
    public void Apply_rejects_unsupported_schema_keywords()
    {
        var schema = ArgumentSchema.Parse(
            """
            {
              "type": "object",
              "properties": {
                "label": {
                  "type": "string",
                  "pattern": "^[A-Z]+$"
                }
              }
            }
            """);

        var result = ArgumentPatchValidator.Apply(
            schema,
            JsonNode.Parse("""{"label":"Save"}""")!,
            JsonNode.Parse("""{"label":"Discard"}""")!);

        Assert.False(result.Accepted);
        Assert.Equal("args.unsupported_keyword", result.Diagnostics.Single().Code);
        Assert.Equal("Save", result.UpdatedArguments["label"]!.GetValue<string>());
    }
}

File: tests/unit/core/cpp/argument_patch_validator_tests.cpp
#include <catch2/catch_test_macros.hpp>
#include <dearstory/core/argument_patch_validator.hpp>
#include <dearstory/core/argument_schema.hpp>

#include <filesystem>
#include <fstream>
#include <sstream>
#include <string_view>

namespace
{
    nlohmann::json read_vector(std::string_view file_name)
    {
        const auto path = std::filesystem::path(DEARSTORY_REPOSITORY_ROOT) / "tests" / "contract" / "core" / "vectors" / file_name;
        std::ifstream stream(path, std::ios::binary);
        REQUIRE(stream.good());

        std::ostringstream buffer;
        buffer << stream.rdbuf();
        return nlohmann::json::parse(buffer.str());
    }
}

TEST_CASE("argument_patch_validator accepts the shared valid patch vector")
{
    const auto vector = read_vector("patch-valid.string.json");
    auto schema = dearstory::core::argument_schema::parse(vector.at("schema").dump()).value();

    const auto result = dearstory::core::apply_patch(
        schema,
        vector.at("currentArguments"),
        vector.at("patch"));

    REQUIRE(result.accepted);
    REQUIRE(result.diagnostics.empty());
    REQUIRE(result.updated_arguments == vector.at("expected").at("updatedArguments"));
}

TEST_CASE("argument_patch_validator rejects the shared invalid enum vector")
{
    const auto vector = read_vector("patch-invalid.enum.json");
    auto schema = dearstory::core::argument_schema::parse(vector.at("schema").dump()).value();

    const auto result = dearstory::core::apply_patch(
        schema,
        vector.at("currentArguments"),
        vector.at("patch"));

    REQUIRE_FALSE(result.accepted);
    REQUIRE(result.diagnostics.size() == 1U);
    REQUIRE(result.diagnostics.front().code == "args.enum");
    REQUIRE(result.updated_arguments == vector.at("expected").at("updatedArguments"));
}

TEST_CASE("argument_patch_validator rejects unsupported schema keywords")
{
    auto schema = dearstory::core::argument_schema::parse(R"({
        "type":"object",
        "properties":{"label":{"type":"string","pattern":"^[A-Z]+$"}}
    })").value();

    auto result = dearstory::core::apply_patch(
        schema,
        nlohmann::json::parse(R"({"label":"Save"})"),
        nlohmann::json::parse(R"({"label":"Discard"})"));

    REQUIRE_FALSE(result.accepted);
    REQUIRE(result.diagnostics.size() == 1U);
    REQUIRE(result.diagnostics.front().code == "args.unsupported_keyword");
    REQUIRE(result.updated_arguments == nlohmann::json::parse(R"({"label":"Save"})"));
}
